
```{r}
library(dplyr)
setwd('/scratch/prj/bioresource/Public/PGx/data')
c19.pos <- read.table('rsid_pos_CYP2C19.txt', header=F) %>% rename(POS=V1)
c19.rsid <- read.table('rsid_CYP2C19.txt', header=F) %>% rename(RSID=V1) %>% cbind(.,c19.pos)
vcf<-read.table(gzfile('GLADv2_imputed_rsid_CYP2C19.vcf.gz'), stringsAsFactors = FALSE)[1:3] %>% rename(CHROM=V1,POS=V2,ID=V3)
c19.vcf <- vcf %>% merge(.,c19.rsid, by='POS') 
```

# 7 SNPs overlap: rs12248560, rs28399504, rs12769205, rs41291556, rs58973490, rs4244285, rs3758581



# check which SNPs belong to one star allele in CYP2C19 using CPIC reference table
```{r}
# rs12248560   *17
# rs12769205   *2 *35
# rs3758581    *1 
# rs41291556   *8
# rs4244285    *2
# rs58973490   *11
# rs28399504   *4
# *2  rs12769205 A>G rs4244285 G>A rs3758581 A>G    No function
# *4  rs28399504 A>G rs3758581 A>G                  No
# *8  rs41291556 T>C                                No
# *11 rs58973490 G>A rs3758581 A>G                  Normal
# *17 rs12248560 C>T rs3758581 A>G                  Increased
# *35 rs12769205 A>G rs3758581 A>G                  No (subset of *2)

```

# manually call CYP2C19 from genotype data
```{r}
# only included overlapped snps (7)
# snps <- c("rs12248560", "rs12769205", "rs3758581","rs41291556", "rs4244285", "rs58973490", "rs28399504")

vcf<-read.table(gzfile('GLADv2_imputed_rsid_CYP2C19.vcf.gz'), stringsAsFactors = FALSE) %>% filter(V2 %in% c19.vcf$POS) %>% merge(c19.vcf[,c('POS','RSID')], ., by.x='POS', by.y='V2')
colnames(vcf) <- header 

write.table(vcf, 'CYP2C19_select.vcf', quote=F, row.names=F)
# GLAD_ref GLAD_alt Pharmvar_ref Pharmvar_alt
# rs12248560    C T     same
# rs12769205    A G     same
# rs3758581     A G     same 
# rs41291556    T C     same
# rs4244285     G A     same
# rs58973490    G A     same
# rs28399504    A G     same

## after finished haplotype and phenotype calling, interestingly, I found all individuals carrying *11 variants (rs58973490) are phased with *2 variants so *11 carriers were replaced by *2 carriers and are not called *11 in the haplotype results.

vcf <- vcf %>% t(.) %>% as.data.frame() %>% .[c(2,5,6,11:dim(.)[1]),]
colnames(vcf) <- vcf[1,]
vcf <- vcf[-1,]
# only leave genotype from vcf
vcf <- as.data.frame(apply(vcf, 2, function(x) {do.call(rbind, strsplit(x,':'))[,1]}))

# separate diplotype to haplotype, the star alleles are ordered from the lowest activity score to the highest
vcf <- vcf %>% mutate(hap_1='*1', hap_2='*1')
# because rs3758581 was covered in all cohorts so I don't use if.
# *2 rs12769205 rs4244285 rs3758581
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs12769205, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs4244285, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,1] =='1' & vcf$hap_1 =='*1')] <- '*2'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs12769205, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs4244285, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,2] =='1' & vcf$hap_2 =='*1')]  <- '*2'

# *4 rs28399504 rs3758581 
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs28399504, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,1] =='1' & vcf$hap_1 =='*1')] <- '*4'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs28399504, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,2] =='1' & vcf$hap_2 =='*1')] <- '*4'

# *8 rs41291556 
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs41291556, '\\|'))[,1] =='1' & vcf$hap_1 =='*1')] <- '*8'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs41291556, '\\|'))[,2] =='1' & vcf$hap_2 =='*1')] <- '*8'

# *35 rs12769205 rs3758581
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs12769205, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,1] =='1'& vcf$hap_1 =='*1')] <- '*35'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs12769205, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,2] =='1'& vcf$hap_2 =='*1')] <- '*35'

# *17 rs12248560 rs3758581
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs12248560, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,1] =='1' & vcf$hap_1 =='*1')] <- '*17'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs12248560, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,2] =='1' & vcf$hap_2 =='*1')] <- '*17'

# *11 rs58973490 rs3758581
vcf$hap_1[which(do.call(rbind, strsplit(vcf$rs58973490, '\\|'))[,1] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,1] =='1'& vcf$hap_1 =='*1')] <- '*11'
vcf$hap_2[which(do.call(rbind, strsplit(vcf$rs58973490, '\\|'))[,2] =='1' & do.call(rbind, strsplit(vcf$rs3758581, '\\|'))[,2] =='1'& vcf$hap_2 =='*1')] <- '*11'

# activity score: this activity score is only used to categorize phenotype easily. it doesn't have any reference support.
vcf <- vcf%>% mutate(as_hap1 = ifelse(
    grepl('^\\*2$|^\\*4$|^\\*8$|^\\*35$', hap_1), 0, ifelse(
        grepl('^\\*17$', hap_1), 1.5, 1
    )))
vcf <- vcf%>% mutate(as_hap2 = ifelse(
    grepl('^\\*2$|^\\*4$|^\\*8$|^\\*35$', hap_2), 0, ifelse(
        grepl('^\\*17$', hap_2), 1.5, 1
    )))
# activity score
vcf <- vcf %>% mutate(as=as_hap1 + as_hap2, diplotype = paste0(hap_1, '/', hap_2))
# phenotype
vcf <- vcf %>% mutate(phenotype = ifelse(
    as>=2.5, 'Ultrarapid', ifelse(
        as==0, 'Poor', ifelse(
            as<=1.5 & as>0, 'Intermediate', 'Normal'
        )
    )))
vcf <- vcf %>% .[3:dim(vcf)[1],]
table(vcf$phenotype)
# Intermediate       Normal         Poor   Ultrarapid 
#         4585         6901          409         5613


# match sample ID
# get header from vcf
system(paste0("zcat /scratch/prj/bioresource/Public/GLADv2/v2_imputed_data/vcf/GLAD_rsid_chr10_maf1_R2_3.vcf.gz | awk '$1 ~ /^#CHROM/ {print NR}'")) # 25

header <- readLines(pipe(paste0("zcat /scratch/prj/bioresource/Public/GLADv2/v2_imputed_data/vcf/GLAD_rsid_chr10_maf1_R2_3.vcf.gz | head -n 25 | tail -n 1"))) 
header <-unlist(strsplit(header[length(header)],"\t")) 
header <- header[10:length(header)]
vcf <- vcf %>% tibble::add_column(ID=header, .before='rs12248560')

vcf %>% saveRDS(file='CYP2C19_starallele_call.rds')
```